src/Evolution/DiscontinuousGalerkin/Actions/NormalCovectorAndMagnitude.hpp

```diff
@@ -0,0 +1,181 @@
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        unnormalized_normal_covector) noexcept {
+  if constexpr (has_inverse_spatial_metric_tag_v<System>) {
+    using inverse_spatial_metric_tag =
+        typename System::inverse_spatial_metric_tag;
+    auto& normal_vector = get<NormalVector<Dim>>(*fields_on_face);
+    const auto& inverse_spatial_metric =
+        get<inverse_spatial_metric_tag>(*fields_on_face);
+    // Compute unnormalized normal vector
+    for (size_t i = 0; i < Dim; ++i) {
+      normal_vector.get(i) = inverse_spatial_metric.get(i, 0) *
+                             get<0>(unnormalized_normal_covector);
+      for (size_t j = 1; j < Dim; ++j) {
+        normal_vector.get(i) += inverse_spatial_metric.get(i, j) *
+                                unnormalized_normal_covector.get(j);
+      }
+    }
+    // Perform normalization
+    dot_product(face_normal_magnitude, normal_vector,
+                unnormalized_normal_covector);
+    get(*face_normal_magnitude) = sqrt(get(*face_normal_magnitude));
+    auto& one_over_normal_vector_magnitude =
+        get<OneOverNormalVectorMagnitude>(*fields_on_face);
+    get(one_over_normal_vector_magnitude) = 1.0 / get(*face_normal_magnitude);
+    for (size_t i = 0; i < Dim; ++i) {
+      unit_normal_covector->get(i) = unnormalized_normal_covector.get(i) *
+                                     get(one_over_normal_vector_magnitude);
+      normal_vector.get(i) *= get(one_over_normal_vector_magnitude);
+    }
+  } else {
+    magnitude(face_normal_magnitude, unnormalized_normal_covector);
+    auto& one_over_normal_vector_magnitude =
+        get<OneOverNormalVectorMagnitude>(*fields_on_face);
+    get(one_over_normal_vector_magnitude) = 1.0 / get(*face_normal_magnitude);
+    for (size_t i = 0; i < Dim; ++i) {
+      unit_normal_covector->get(i) = unnormalized_normal_covector.get(i) *
+                                     get(one_over_normal_vector_magnitude);
+    }
+  }
+}
+
+/*!
+ * \brief Computes the normal vector, covector, and their magnitude on the face
+ * in the given direction.
+ */
+template <typename System, size_t Dim, typename DbTagsList,
+          typename FieldsOnFaceTags>
+void unit_normal_vector_and_covector_and_magnitude(
+    const gsl::not_null<db::DataBox<DbTagsList>*> box,
+    const gsl::not_null<Variables<FieldsOnFaceTags>*> fields_on_face,
+    const Direction<Dim>& direction,
+    const std::unordered_map<Direction<Dim>, tnsr::i<DataVector, Dim>>&
+        unnormalized_normal_covectors) noexcept {
+  const bool mesh_is_moving =
+      not db::get<domain::CoordinateMaps::Tags::CoordinateMap<Dim, Frame::Grid,
+                                                              Frame::Inertial>>(
+              *box)
+              .is_identity();
+  const auto& unnormalized_normal_covector =
+      unnormalized_normal_covectors.at(direction);
+
+  db::mutate<
+      evolution::dg::Tags::InternalFace::NormalCovectorAndMagnitude<Dim>>(
```

Comment:
Seeing as you only mutate the one tag, it seems like it would be more flexible to just take that by not_null argument as well, moving the `db::mutate` into the caller of this free function. I think the only other thing you need from the box is either `CoordinateMap` or just a boolean indicating whether it's using moving mesh?